Pipe Dispenser Improvements/Atmospherics Improvements

Textor44 opened the pull request:
**This PR is not ready and should not be merged.**

Current status:
- [x] Pipe dispenser UI complete.
- [x] Pipe datums are in place.
- [x] Removed non-omni filters and mixers.
- [x] Fixed maps that had non-omni filters and mixers.
- [x] Makes gas sensors buildable and programmable
- [x] Makes injectors buildable and programmable
- [x] Make injectors connect properly when built
- [x] Make injectors rotate properly when clicked on in-hand
- [x] Make injectors toggle when clicked
- [x] Do the new UI for disposal pipe dispensers
- [x] make disposal pipe dispenser pipes part of the pipe datums
- [x] add drains to pipe dispenser
- [x] make disposal pipe segments collide again.
- [x] redo atmos control console UI
- [ ] allow atmos control consoles to talk to sensors built by players
- [ ] allow the multitool to set id tags for equipment and the consoles.
- [ ] make gasses more flexible so that sensors, air alarms, etc. can actually list out unusual gas compositions without having to make specific cases for each one.
- [ ] create hot tapper/bleeder tool.
- [ ] add pipe bag item
- [ ] map in hot tapper and pipe bags
- [ ] slay the evil that is construction.dm
- [ ] more when I think of it.

New Pipe Interface:
![image](https://user-images.githubusercontent.com/12721324/45606177-be907900-b9f7-11e8-8c6d-8b08d835368c.png)

New Atmospherics Console Interface (Prototype)
![image](https://user-images.githubusercontent.com/12721324/45737236-1bc82e00-bba2-11e8-96ea-70c9fc82f1aa.png)

Textor44 committed:
Initial commit

Creates pipe_datums.dm
Modifies pipe_dispenser.dm to override ONLY the pipe and supply pipe to test the new system.

**This PR is not ready and should not be merged.**

PsiOmegaDelta commented on code/game/machinery/pipe/pipe_datums.dm:
```diff
@@ -0,0 +1,73 @@
+/datum/pipe
+	var/name = null				//item's name
+	var/desc = null				//item description
+	var/item_name = null		//original item's name, in case it gets modified.
```
This can be done using `initial(name)` instead

Textor44 replied:
So remove item_name entirely?

PsiOmegaDelta commented on code/game/machinery/pipe/pipe_datums.dm:
```diff
@@ -0,0 +1,73 @@
+/datum/pipe/proc/AssemblePipeInfo()
+	AssemblePipeName()
+	AssemblePipeDesc()
+	return
```
Unnecessary return

PsiOmegaDelta commented on code/game/machinery/pipe/pipe_datums.dm:
```diff
@@ -0,0 +1,73 @@
+		var/atom/movable/A = build_path
+		name = initial(A.name)
+		item_name = name
+	return
```
Unnecessary return

PsiOmegaDelta commented on code/game/machinery/pipe/pipe_datums.dm:
```diff
@@ -0,0 +1,73 @@
+/datum/pipe/proc/AssemblePipeDesc()
+	if(!desc)								//Try to make up a nice description if we don't have one
+		desc = "\a [item_name]."
+	return
```
Unnecessary return

PsiOmegaDelta commented on code/game/machinery/pipe/pipe_dispenser.dm:
```diff
@@ -79,6 +79,19 @@
+		new_item.color = D.pipe_color
+		new_item.name = D.name
+		new_item.desc = D.desc
+	return
```
Unnecessary return

PsiOmegaDelta commented on code/game/machinery/pipe/pipe_datums.dm:
```diff
@@ -0,0 +1,73 @@
+	var/name = null				//item's name
+	var/desc = null				//item description
+	var/item_name = null		//original item's name, in case it gets modified.
+	var/item_id = "id"			//ID for easy reference. All lowercase, alpha numeric, no symbols.
```
Doesn't actually appear to be read anywhere?

Textor44 replied:
Heavily work in progress. If I end up not needing it, it will be removed.

afterthought2 replied:
Moving away from string/number IDs would be good in any case. Current system's biggest issue is the way it relies on numerical IDs. I suggest using type paths as much as possible instead.

PsiOmegaDelta commented on code/game/machinery/pipe/pipe_datums.dm:
```diff
@@ -0,0 +1,73 @@
+	var/desc = null				//item description
+	var/item_name = null		//original item's name, in case it gets modified.
+	var/item_id = "id"			//ID for easy reference. All lowercase, alpha numeric, no symbols.
+	var/build_type = null		//For this set, it's whether it is built at PIPE_DISPENSER or DISPOSAL_DISPENSER
```
A better alternative, have `/datum/pipe/pipe_dispenser` and `/datum/pipe/disposal_dispenser` subtypes

Textor44 replied:
So instead of /datum/pipe/straight I should do /datum/pipe/pipe_dispenser/straight?

PsiOmegaDelta commented on code/game/machinery/pipe/pipe_datums.dm:
```diff
@@ -0,0 +1,73 @@
+	var/sort_string = "ZZZZ"	//Allows for nudging items to where you want them to appear.
+	var/pipe_type = 0			//What sort of pipe this is
+	var/connect_types = 0		//what sort of connection this has
+	var/pipe_color = 0			//what color the pipe should be by default
```
0 doesn't seem like a valid color.

Textor44 replied:
Yeah, I realized that after the fact. I'll set it to a better default.

Textor44 committed:
Fixes Issues, Restructures, Lays groundwork for new menu loading system.

Fixes undef issues with pipe_datums.dm
Removes build_type, replaces it with a value in the path itself.
removes item_name
removes extraneous returns

Textor44 committed:
Removes Return, Corrects issue that may have caused travis failure... hopefully

Textor44 committed:
Adds nanoUI interface to pipe dispenser, eliminates a TON of old code.

Textor44 committed:
Tons of changes

-Adds CONNECT_TYPE_FUEL to definitions
-changes default for build_path to /obj/item/pipe
-adds dir to structure
-adds icon_state to structure
-restructures datums again to allow easier categorization
-rewrites the UI to nanoUI, lists all available datums in the appropriate categories
-changes check-paths.sh to make travis not hate me for removing 4 << things from the old code
-build in pipe_dispenser.dm now checks for invalid values for devices bieng printed (eliminates runtime errors caused by this)

Textor44 commented:
![nanoUI](https://media.discordapp.net/attachments/256853514624434177/484532111014101003/unknown.png)
Opinions welcome.

afterthought2 commented on code/game/machinery/pipe/pipe_dispenser.dm:
```diff
@@ -6,99 +6,93 @@
-
-/obj/machinery/pipedispenser/attack_hand(user as mob)
-	if(..())
+	var/list/simple_recipes = list()
```
How about a structure like `list(name_of_category = list(recipes))` to cut down on the copypaste?

afterthought2 commented on code/game/machinery/pipe/pipe_dispenser.dm:
```diff
@@ -6,99 +6,93 @@
+
+/obj/machinery/pipedispenser/Initialize()
+	. = ..()
+	for(var/simple in typesof(/datum/pipe/pipe_dispenser/simple) - /datum/pipe/pipe_dispenser/simple)
```
`subtypesof`

Textor44 committed:
Made suggested improvements.

Textor44 committed:
adds TONS of new datums, continues UI improvements

-adds most of the missing datums
-adds a color button for regular pipes
-changes the headers from H3 to a colored strong
-adds icon to data structure to accomodate unusual items that are printable
-adds 5x and 10x buttons to dispense 5 or 10 items at a time.

Textor44 commented:
New UI prototype:
![nanoUI](https://media.discordapp.net/attachments/484641323530453003/484691865929515049/unknown.png)

Textor44 committed:
Moves variable from global scope to object scope.

Textor44 committed:
Updating to current version.

afterthought2 commented on code/game/machinery/pipe/pipe_dispenser.dm:
```diff
@@ -6,99 +6,111 @@
+
+/obj/machinery/pipedispenser/Initialize()
+	. = ..()
+	for(var/simple in subtypesof(/datum/pipe/pipe_dispenser/simple))
```
How about something like:
```
/datum/pipe
	var/category = "general recipes"
	var/colorable = FALSE

/datum/pipe/pipe_dispenser/simple
	category = "Regular Pipes"
	colorable = TRUE

...

/obj/machinery/pipedispenser
	...
	var/list/categories = list(
		/datum/pipe/pipe_dispenser/simple,
		/datum/pipe/pipe_dispenser/scrubber,
		/datum/pipe/pipe_dispenser/supply,
		/datum/pipe/pipe_dispenser/fuel,
		/datum/pipe/pipe_dispenser/device
		/datum/pipe/pipe_dispenser/he)
	...

/obj/machinery/pipedispenser/Initialize()
	. = ..()
	for(var/category_type in categories)
		for(var/recipe_type in subtypesof(category_type))
			LAZYADD(categories[category_type], new recipe_type(src))

/obj/machinery/pipedispenser/proc/get_console_data()
	. = list() // There is a performance improvement with doing this. Also, ..() doesn't do anything here.
	for(var/category in categories)
		var/datum/pipe/cat = category // False type
		. += "<table><tr><td><font color = '#517087'><strong>initial(cat.category)</strong></font></td></tr>"
		for(var/datum/pipe/pipe in categories[category])
			var/line = "[pipe.name]</td>"
			. += "<tr><td>[line]<td><a href='?src=\ref[src];build=\ref[pipe]'>Dispense</a></td><td><a href='?src=\ref[src];buildfive=\ref[pipe]'>5x</a></td><td><a href='?src=\ref[src];buildten=\ref[pipe]'>10x</a></td></tr>"
		if(initial(cat.colarable))
			. += "<tr><td>Color</td><td><a href='?src=\ref[src];color=\ref[src]'><font color = '[SimplePipeColor]'>[SimplePipeColor]</font></a></td></tr>"
	. = JOINTEXT(.)
```

Textor44 committed:
Merge remote-tracking branch 'upstream/dev' into Atmospherics-Improvements

Textor44 committed:
Begun rewriting construction.dm

-

-Added white as a color for pipes, replacing the "grey" (that was actually white) and giving it the standard grey color.
-Rewrote /obj/item/pipe/New() to be far smaller and more flexible.
-Added new variables to /obj/machinery/atmospherics/
-changed icon and icon_state to build_icon and build_icon_state to match variables in the /obj/machinery/atmospherics definitions for better compatibility with  /obj/item/pipe/New()
-moved build function to /datum/pipe
-adjusted most /obj/machinery/atmospheric items to include the new variables.

Hopefully these changes aren't too horrifying...

Textor44 committed:
Merge remote-tracking branch 'upstream/dev' into Atmospherics-Improvements

Textor44 committed:
Murders non-omni filters and mixers

-Excises the non-omni filters and mixers from the code entirely.
-Replaces the last remaining old-style mixer in the incinerator with a proper omni-filter.
-Redefines pipe values so they are more extensible and easier to follow.

Textor44 committed:
Purges outdated filters and mixers from away sites.

-Removes filters and mixers from Errant Pisces, replacing them with omni variants
-Removes filters and mixers from icarus, replacing them with omni variants
-Removes filters and mixers from lar maria, replacing them wtih omni variants.

PsiOmegaDelta commented on code/game/machinery/pipe/pipe_datums.dm:
```diff
@@ -0,0 +1,422 @@
+/datum/pipe
+	var/name = "pipe"							//item's name
+	var/desc = "a pipe"							//item description
+	var/item_name = null						//original item's name, in case it gets modified.
```
Looking through the code I admit I fail to see the purpose of this particular var, even with the comment.

Textor44 replied:
It'll likely get axed. I modeled it after the protolathe code.

PsiOmegaDelta commented on code/game/machinery/pipe/pipe_datums.dm:
```diff
@@ -0,0 +1,422 @@
+
+/datum/pipe/proc/Build(var/datum/pipe/D, var/PipeColor = PIPE_COLOR_GREY, var/loc)
+	if(D.build_path)
+		var/obj/item/pipe/new_item = D.Fabricate(src)
```
..why just not supply `loc` here directly instead of creating the pipe at one location then immediately relocating it to another?

PsiOmegaDelta commented on code/__defines/atmospherics.dm:
```diff
@@ -1,63 +1,69 @@
+#define PIPE_OMNI_FILTER			610
+#define PIPE_UNIVERSAL				611
+#define PIPE_MTVALVEM				612
+#define PIPE_SVALVE					613
```
Do all these defines serve a purpose anymore?

Textor44 replied:
They are used all over atmos calls. I'd rather have them make more sense on the back end than try to interpret all the code with pure numbers.

PsiOmegaDelta commented on code/game/machinery/pipe/pipe_dispenser.dm:
```diff
@@ -6,99 +6,108 @@
+	for(var/datum/pipe/devices in categories["device_recipes"])
+		var/line = "[devices.name]</td>"
+		result += "<tr><td>[line]<td><a href='?src=\ref[src];build=\ref[devices]'>Dispense</a></td><td><a href='?src=\ref[src];buildfive=\ref[devices]'>5x</a></td><td><a href='?src=\ref[src];buildten=\ref[devices]'>10x</a></td></tr>"
+	. += "[result]"*/
```
Don't keep dead code.

PsiOmegaDelta commented on code/game/machinery/pipe/pipe_dispenser.dm:
```diff
@@ -6,99 +6,108 @@
+		if(!choice)
+			return
+		pipe_color = choice
+		updateUsrDialog()
```
Handles HREFs should return 1

Textor44 committed:
Removes filters and mixers from slavers base and casino

-Removes outdated filters and mixers from slavers base
-removes outdated filters and mixers from casino

Textor44 committed:
Makes requested changes, improves away site replacements.

-Removes unnecessary variable from pipe datum
-Purges obsolete code
-Makes Errant Pisces and Icarus maps have their omni-x devices setup like they should be.

Textor44 committed:
Starts minor work on atmos control consoles.

Textor44 commented:
This is definitely still a WIP. While reviews are welcome, there will still be yet more changes coming and the code will keep being modified for a while.

Textor44 committed:
Merge remote-tracking branch 'upstream/dev' into Atmospherics-Improvements

Textor44 committed:
Fixes Check-paths.sh again

makes it so travis stops hating the PR for removing things.

Textor44 committed:
Adds gas sensors and injectors to the dispenser.

Textor44 committed:
Merge remote-tracking branch 'upstream/dev' into Atmospherics-Improvements

Textor44 committed:
Injector construction improvements.

-Fixes injector building directions
-Allows injectors to be rotated in hand
-Allows injectors to be rotated by alt-clicking
-Automatically builds pipe network when built.
-Fixes issue with injectors always being built in the same direction.
-Fixes runtime when building an injector at the dispenser.

Textor44 committed:
Makes Construction.dm slightly less terrible.

-Makes four new procs that handle connecting pipes to networks to reduce repeated code
-Fixes issue with unwrenched manifolds turning into 4w manifolds after wrenching them back down.

Textor44 committed:
Begins work on disposals

-Currently broken. Extremely broken.
-Replaces numeric references with defines in order to make the code easier to follow
-Breaks disposal.dm so that disposals can't be used anymore. This isn't a feature, though.

Textor44 committed:
Fixes disposals

-Disposals are no longer broken.
-Splits Disposal.dm into three files so that they are more managable and things are easier to find (hopefully)

Textor44 committed:
Finishes disposal pipe dispenser improvement

-Disposal pipe dispenser now dispenses all disposal pipe types
-Disposal pipe dispsener uses the new UI style.
-Removes base_icon thing from disposal-construction.dm since it added an extra layer of complexity with no benefit.

Textor44 commented:
Disposal pipe UI:
![image](https://user-images.githubusercontent.com/12721324/45605838-285b5380-b9f5-11e8-86c8-d8219ce96cdf.png)

Textor44 committed:
Adds drains

- Adds drains to pipe dispensers.
-Creates new drain item for construction purposes.

Textor44 committed:
Makes travis love me again

Textor44 committed:
Merge remote-tracking branch 'upstream/dev' into Atmospherics-Improvements

Textor44 committed:
Merge remote-tracking branch 'upstream/dev' into Atmospherics-Improvements

Textor44 committed:
Updates density on pipes from disposal dispenser

Textor44 committed:
Adjusts atmos control UI

-Changes Atmos Control to NanoUI
-Moves supermatter control console under the large_tank_control subtype
-Adds direct input for pressure/flow rate
-Adds max buttons for pressure/flow rate
-Updates map with supermatter control console path update.
-updates other paths for change.

PsiOmegaDelta commented on code/game/machinery/atmo_control.dm:
```diff
@@ -77,12 +88,47 @@ obj/machinery/air_sensor/Destroy()
+/obj/machinery/air_sensor/attackby(var/obj/item/O as obj, var/mob/user as mob)
+	if(isMultitool(O))
+		var/t = sanitizeSafe(input(user, "Enter the ID for the sensor.", src.name, id_tag), MAX_NAME_LEN)
+		if(user.incapacitated() && !user.Adjacent(src))
```
`CanPhysicallyInteract(user)`

PsiOmegaDelta commented on code/game/machinery/atmo_control.dm:
```diff
@@ -77,12 +88,47 @@ obj/machinery/air_sensor/Destroy()
+		if (t)
+			src.id_tag = t
+			to_chat(user, "<span class='notice'>The new ID of the sensor is [id_tag]</span>")
+		return
```
Can use `if`/`else if` instead of returns

PsiOmegaDelta commented on code/game/machinery/atmo_control.dm:
```diff
@@ -444,7 +506,7 @@ Min Core Pressure: [pressure_limit] kPa<BR>"}
 	..()
 
-/obj/machinery/computer/general_air_control/fuel_injection/return_text()
+/*/obj/machinery/computer/general_air_control/fuel_injection/return_text()
```
Making a note: Don't keep dead code.

PsiOmegaDelta commented on code/game/machinery/pipe/pipe_datums.dm:
```diff
@@ -0,0 +1,610 @@
+
+/datum/pipe/proc/AssemblePipeDesc()
+	if(!desc)								//Try to make up a nice description if we don't have one
+		desc = "\a [name]."
```
Descriptions should begin with an uppercase.
`"\A [name]"`

Textor44 committed:
Makes requested changes

-Makes requested changes
-Fixes univeristy ship map.

Textor44 committed:
Merge remote-tracking branch 'upstream/dev' into
Atmospherics-Improvements

Textor44 committed:
Makes atmos scanners more precise

-Increases the resolution of atmos scanners
-Adds moles to the data given in advanced mode.

Roland410 commented:
Not sure if it's out of the scope of the PR, but what about porting /tg/'s RPD's if you're able?

Textor44 commented:
@Roland410 what's /tg/'s RPD?

Roland410 commented:
Rapid pipe dispenser, atmos gets two, one in each locker. It's basically a
portable pipe dispenser which can color, destroy etc. pipes including
disposal etc.

On Fri, Sep 21, 2018, 19:29 James Andersen <notifications@github.com> wrote:

> @Roland410 <https://github.com/Roland410> what's /tg/'s RPD?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Baystation12/Baystation12/pull/22880#issuecomment-423610486>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGRsuxU5QQ2OK_4RiCayGVH_nA-uC3dhks5udR9ZgaJpZM4WRaa->
> .
>

babydoll commented:
Please let me know when you're ready to resume work on this.
